docs/architecture-core.md

```diff
@@ -0,0 +1,174 @@
+in a similar fashion to the in-kernel `overlayfs`.
+(See also https://github.com/ostreedev/ostree/issues/2281)
+
+### Kernel handling
+
+ostree is entirely oriented around bootable filesystem trees;
+its "source of truth" is the bootloader entries.  It has opinions
+about where the Linux kernel binaries are stored (the current
+standard is in `/usr/lib/modules/$kver`.)
+
+In contrast, traditional RPM is unaware of what a kernel is; it's
+just another package.  Most higher level package managers such
+as yum gained some special casing around the kernel - because
+it's not possible to restart the running kernel, traditional RPM
+systems need to keep the kernel modules for the running kernel
+around.  For example yum/dnf have a concept of "installonlyn"
+which defaults to 2 for the kernel package.
+
+Additionally, for at least traditional Fedora derivatives with
+yum/dnf, the initramfs is generated client side as part of
+a kernel update.
+
+But for rpm-ostree, the decision was made to default to a
+pre-generated initramfs by default.  Further, in order to implement transactional
+upgrades, rpm-ostree needs to be in control of the initramfs
+regeneration - it can't just be a script forked off without its
+knowledge.
+
+Further for rpm-ostree, easily replacing the kernel (as well as userspace)
+is intended to be a first-class operation; you need to be able to do that
+in order to debug production issues for example.
+
+In contrast to the yum/dnf "installonly" for ostree there can be exactly one kernel per userspace
+filesystem tree.  To ostree, a "bootable ostree commit"
+is the pair of (kernel, userspace).
+
+rpm-ostree combines these two worlds, and goes to some
+lengths to bend the libdnf stack to work this way.  We reset
+the "installonly" limit back to 1 to ensure we have exactly
+one kernel.  PR: https://github.com/coreos/rpm-ostree/pull/1228
+
+Further, as noted above rpm-ostree takes over the handling of
+invoking `dracut` - just like other scripts, it is run inside
+a container with just read-only access to the system.  `dracut`
+generates the initramfs CPIO archive, which we then place inside
+the `/usr/lib/modules/$kver` location.
+
+If client-side initramfs regeneration is enabled, we may selectively
+provide desired configuration files into this process.  PR: https://github.com/coreos/rpm-ostree/pull/2170
+
+### SELinux
+
+Handling SELinux is very tricky, because it is a package that can affect
+*every other package*.  Specifically, the SELinux policy package
+contains a vast set of regular expressions in `file_contexts`
+to determine labeling.
+
+For traditional librpm, this is a plugin.
+
+A major goal of OSTree from the start has been to ensure fully correct
+handling of SELinux for the base operating system.  The way
+rpm-ostree handles this is by loading the policy after the target
+filesystem tree has been generated, and passes that loaded policy
+to libostree, which consults it to use for the label of each
+committed file.
```

Comment:
Probably also worth mentioning that at compose time we rebuild the policy before committing (with that fresh policy).